Arelle/Arelle
# Fix described errors in xbrlDB plugin

## Oracle database connection
#### Reason for change
Error connecting to oracle database  
#### Description of change
1374359b84f498794b1cf9b6a6e026da251426a8 changes include:
1. Add database name to oracle connection string.
2. Correct default port for mysql.
3. Correct table name format for lock table statement.
#### Steps to test
Connect to Oracle database (requires user has connect privileges).
```python
from arelle.plugin.xbrlDB.XbrlSemanticSqlDB import XbrlSqlDatabaseConnection
from arelle.Cntlr import Cntlr
con = XbrlSqlDatabaseConnection(Cntlr(), user='<user name>', password='<password>', host='<host>', port='<port default 1521>', database='<database name>', timeout=None, product='orcl')
con.execute('SELECT * FROM V$VERSION')
```
Also implicitly tested if Oracle connection is provided for unit tests in ~/tests/unit_tests/arelle/plugin/xbrlDB/xbrldb_test_helpers.py:conn_params

## Create Schema
#### Reason For Change
Errors when automatically initializing database schema with Oracle, MySql and MSSQL
#### Description of change
b1903533ec7d7be9df0191710acfd4c8ecee9b6d includes the following changes:
1. Avoid database product related errors when parsing SQL statements from ddl files.
2. Fix `sequencesInDB` method to get correct objects based on database product.
3. Add  `triggersInDB` method to get triggers based on product.
4. Disable default `dropPriorTables`, because this drops ALL tables and objects in database not only objects created by this plugin.
5. Add step to drop triggers when `dropPriorTables` is selected.
#### Steps to Test
Unit test -> ~/tests/unit_tests/arelle/plugin/xbrlDB/test_xbrlDB_semantic_sql.py:test_create
## Automatic schema initialization
#### Reason for change
Database schema is initialized automatically, while according to arelle documentation, users should manually initialize schema using the ddl files (see [SQL Schema Initialization](https://arelle.org/arelle/documentation/xbrl-database/)).
#### Description of change
293ddf6bc241f3199f4f7bb260bf13e849a37609 Disable automatic initialization of schema, and provide user message with appropriate ddl file to initialize the schema.  
Note: this disables automatically triggering schema initialization only, `create` method is still usable and is used in unit test.
#### Steps to Test
Unit test -> ~/tests/unit_tests/arelle/plugin/xbrlDB/test_xbrlDB_semantic_sql.py:test_disable_automatic_create
## Oracle DDL file
#### Reason for change
Missing sequence and invalid data types
#### Description of change
8d9345c2e3e7519714296cd36fadc46b1a826e9e Add create statement for missing sequence and correct data types.
#### Steps to Test
Run file `arelle/plugin/xbrlDB/sql/semantic/xbrlSemanticOracleDB.sql` it should run without errors and create all objects. Also implicitly tested if Oracle connection is provided for unit tests in ~/tests/unit_tests/arelle/plugin/xbrlDB/xbrldb_test_helpers.py:conn_params

## NAICS industry codes industry codes 31-33, 44-45, 48-49 (optional)
#### Reason for change
NAICS industry codes industry codes 31-33, 44-45, 48-49 entered in `industry` table as having depth 4 instead of 1 (see [census.gov/naics > Reference Files > Downloadable Files > 2017 NAICS Descriptions [XLSX, 264KB]](https://www.census.gov/naics/?48967)), and table `industry_level` is missing levels for these codes.  
Note: this is optional data, does not have any effect on functionality.
#### Description of change
116e8ef96f1596924c145caca75f468b2ab5fc66 Changes include:
1. Entered correct depth for NAICS industry codes 31-33, 44-45, 48-49 in table `industry` data.
2. Added `industry_level` rows for the mentioned industry codes (see details [here](https://gist.github.com/selgamal/d0aa4c31633b86b7aa963f708ad2f074)).
3. Misc typos in `industry` table data in file `xbrlSemanticSQLiteDB.ddl`.
#### Steps to Test
Review changes to files and the generation of additional rows detailed [here](https://gist.github.com/selgamal/d0aa4c31633b86b7aa963f708ad2f074).

## KeyError on inserting into DB (in specific cases)
#### Reason for change
1. Inserting a filing having `duplicatedSchema warning` produces `KeyError`. The reason for this error is:
   1. When there is a duplicated namespace (2 urls with schemas targeting same namespace) arelle creates only one modelDocument and assigns it to both urls in `ModelXbrl.urlDocs`, so there are 2 keys with same modelDocument as value in the `ModelXbrl.urlDocs` dict.
   2. In `XbrlSqlDatabaseConnection.identifyPreexistingDocuments()` connection attribute `self.urlDocs` dict is used to store documents by url, the url key is based on `modelDocument.uri` so one of the urls will be missing, which results in a `KeyError` later in `XbrlSqlDatabaseConnection.insertDocument()` when `self.documentIds` is created.
   3. Example filings with this issue from ESEF filling index [here](https://filings.xbrl.org/959800H2P9S8MS95DT42/2021-12-31/ESEF/ES/0/), and [here](https://filings.xbrl.org/95980079E2NBJT967T79/2021-12-31/ESEF/ES/0/). This error can be triggered by running (before this pr):
   ```shell
   arelleCmdLine -f https://filings.xbrl.org/95980079E2NBJT967T79/2021-12-31/ESEF/ES/0/95980079E2NBJT967T79-20211231-ES.zip --plugin 'inlineXbrlDocumentSet|xbrlDB' --logFile logToPrint --store-to-XBRL-DB ',,,,:memory,,sqliteSemantic'
   ```
2. When loading a filing from zip archive, `ModelXbrl.modelDocument` is of type `INLINEXBRLDOCUMENTSET` which is not included in document types loaded into the database, while used as document identifier for in-instance relationship sets when inserting data into table `relationship_sets` (such as footnotes). This error can be triggered by running:
```shell
# instance should have footnotes to triggered this error
arelleCmdLine -f https://www.sec.gov/Archives/edgar/data/789019/000156459021020891/0001564590-21-020891-xbrl.zip --plugin 'inlineXbrlDocumentSet|validate/EFM|transforms/SEC|xbrlDB' --logFile logToPrint --store-to-XBRL-DB ',,,,:memory:,,sqliteSemantic'
```
#### Description of change
82ad66cbbf0f91d1e7077647c2a0b8c931109323 Includes the following changes:
1. Include `Type.INLINEXBRLDOCUMENTSET` in `isSemanticDocument` method.
2. When inserting `urlDocs` Use `ModelDocument.uri` instead of url to avoid keyError.
#### Steps to Test
Unit test -> ~/tests/unit_tests/arelle/plugin/xbrlDB/test_xbrlDB_semantic_sql.py:test_insert_xbrl

## Pre-existing documents are inserted due to url http/https difference
#### Reason for change
When inserting a filing `xbrlDB` plugin identifies filing documents that are already in the database (standard schemas and linkbases), documents are identified based on url, and when the same document has a different url due to protocol change from http to https, documents are re-inserted in database.
#### Description of change
e2aba70ac44e96243c7224f20768f3d8f65f3a81 Changed query to detect existing documents to compare url excluding protocol.
#### Steps to Test
Unit test -> ~/tests/unit_tests/arelle/plugin/xbrlDB/test_xbrlDB_semantic_sql.py:test_detect_preexisting_docs

## Add test for xbrlDB plugin
Tests run by default on sqlite in memory database, but other database products can be tested if connection parameters are added to `con_params` dict in file `~/tests/unit_tests/arelle/plugin/xbrlDB`.

Also filings used for tests can be added to `test_filings` list currently containing only 2 problematic filings.

I ran tests locally using sqlite, postgres, mysql, oracle, mssql and all were successful, used docker containers for mysql, oracle, and mssql and here are some notes:
1. MySql was VERY slow compared to all others, it takes about 40 mins to complete the tests while other products were in the range of 3-5 mins (tested before and after pr).
2. Had to install driver for MSSQL and change its connection string temporarily while testing to use driver `DRIVER={{ODBC Driver 17 for SQL Server}}`.

**review**:
@Arelle/arelle
